src/js/program/program-controller.js

```diff
@@ -321,48 +351,62 @@ class ProgramController extends Eventable {
      * @private
      */
     _destroyActiveMedia() {
-        const { mediaPool, mediaController, model } = this;
+        const { mediaController, model } = this;
         if (!mediaController) {
             return;
         }
 
         mediaController.detach();
-        mediaPool.recycle(mediaController.mediaElement);
-        mediaController.destroy();
-
+        this._destroyMediaController(mediaController);
         model.resetProvider();
         this.mediaController = null;
     }
 
     /**
-     * Destroys background media.
+     * Destroys all background media.
      * @returns {void}
      * @private
      */
     _destroyBackgroundMedia() {
-        const { backgroundMedia, mediaPool } = this;
-        if (!backgroundMedia) {
+        const { background } = this;
+        this._destroyMediaController(background.currentMedia);
+        background.currentMedia = null;
+        this._destroyBackgroundLoadingMedia();
+    }
+
+    /**
+     * Destroys a mediaController, and returns it's tag to the pool.
+     * @param {MediaController} mediaController - The media controller to destroy and recycle.
+     * @returns {void}
+     * @private
+     */
+    _destroyMediaController(mediaController) {
+        const { mediaPool } = this;
+        if (!mediaController) {
             return;
         }
-        mediaPool.recycle(backgroundMedia.mediaElement);
-        backgroundMedia.destroy();
-        this.backgroundMedia = null;
+        mediaPool.recycle(mediaController.mediaElement);
+        mediaController.destroy();
     }
 
     /**
-     * Loads the constructor required for the current source.
-     * Resolves with the required constructor, or rejects when the constructor could not be found or loaded.
-     * If rejected, current playback will be destroyed.
+     * Constructs a new media controller with the provider whose able to play the current source.
+     * Will wait and load the provider constructor if it has not already been loaded.
+     * If the required provider cannot be loaded, the subsequent promise rejection will destroy playback.
      * @param {Source} source - The playlist item Source for which a provider is needed.
      * @returns {Promise} The Provider constructor promise.
      * @private
      */
-    _loadProviderConstructor(source) {
+    _setupMediaController(source) {
         const { model, providerController } = this;
+        const makeMediaController = ProviderConstructor => new MediaController(
+            new ProviderConstructor(model.get('id'), model.getConfiguration(), this.primedElement),
+            model
+        );
 
         let ProviderConstructor = providerController.choose(source);
         if (ProviderConstructor) {
-            return Promise.resolve(ProviderConstructor);
+            return Promise.resolve(makeMediaController((ProviderConstructor)));
         }
 
         return providerController.loadProviders(model.get('playlist'))
```

Comment:
We should consider changing `loadProviders` to only load the provider needed for `source`.

Reply:
I would be great to change or replace this method with one that accepts `source` and async returns `ProviderConstructor `.